Connect session and envelope.

kozzztik opened the pull request:
We can avoid using many parameters for calls by connecting Envelope and Session together, and Session with protocol.
Also, sure, that STARTTLS handler needs only session info (anyway, envelope and protocol still avaliable through session).

warsaw commented:
Questions:

In this branch, the session keeps a reference to the envelope, and the envelope keeps a reference to the session.  Are you concerned about the circular reference this sets up?

Since there is now this circular reference, why do handler hooks take an envelope instance instead of a session instance?

Do you think that having this circular reference makes the SMTP class a little more fragile?  I note that `_set_rset_state()` had to be called in a different order in `connection_made()`.  The reason for this is that the former calls `_set_post_data_state()` and that creates a new envelope and sets it on the session object, so that has to be called *after* the new session is created in `connection_made()`.

I notice that the SMTP instance also still keeps a reference to the envelope.  What should the policy/advice be for subclasses of SMTP?  Should they use `self.envelope` or `self.session.envelope`?  Is there any possibility that the two will get out of sync?   Shouldn't there be only one way of doing it?

Ultimate the question is whether this makes writing handlers easier to understand and more readable.  I'll note that there are no documentation changes in this branch, so it's not clear to me whether the explanation of how to do this is better or worse under this change.  It made a lot of sense to me to separate the envelope and session, because that was easy to explain, and the two objects had different lifetimes.  But by conflating the two instances, the question is whether that confuses the matter or makes it clearer.  I'm not sure.  One thing to mention is that for *me*, the length of the signatures of the handler hooks is not a probably when the two concepts are separated.  In the old branch it was easy to explain what the arguments `server`, `session`, and `envelope` represented, and where to find the information you cared about in your handler.

Since this change is an API change, I want to make sure we get it right for 1.0 final.  I may have to defer bug fix changes until after 1.0 if we're going to get that out before Pycon.

kozzztik commented:
Yep, that may be a little bit confusing. Idea to connect session to envelope was to add additional shorthand. But looks like it if more confusing than useful. I removed it.
Docs updated.
About _rset order: I changed it one more time. That is not very obvious, but _rset have it is own context. First one - it is made in Session context, thats why I moved it after session create. Second one - is it is related to handler task. Session creation is connected with loop socket accept, so it lives inside special connection handler. But RSET command itself run within _handle_client coroutine task, end all envelope changes are strongly related to it. I found that during my integrations. I use some workaround to provide context informations to logging system using stdlib logging module. But it don`t know anything about loop, so the only way to provide context information is to use current loop current task object. When envelope is created it my project, it generates unique id, that is context related to any message in logs. That is very useful to debug, as I can separate SMTP session logging messages per connection. Also, message id is returned with DATA command in postix-like way, so I can trace my message in system. If something goes wrong, I request that message id from request side, and then query DB for all logs.
Only problem here was that despite envelope strongly lives within coroutine handling task, its creation happens before task is created. I changed this behavior by moving _rset to task context.

warsaw commented:
What do you think about splitting out the moving of `_set_rset_state()` into a separate PR, with some separate tests?   I'm still not sure about combining the envelope and session at all, i.e. whether that improves understanding, reasoning, the ability to explain and document, etc.  But `_set_rset_state()` seems independent of that, so it would be better to handle that in a separate PR.

(I'm also sorry, but I'm not sure I completely understand your description above.  But I think we can discuss that in a separate PR too.)

kozzztik commented:
Ok, I removed it. Will made it separete PR later.

warsaw commented:
Interestingly enough, I had an opportunity to describe the inner workings to someone recently, and it was really nice to be able to explain the difference between the envelope and the session.  The person I was explaining it to was able to immediately understand the different lifetimes as they relate to the sequence of events that a client goes through, e.g. connect, HELO, RSET.  So I think I want to keep the separation between these two instances, and not combine them.

Given the conflicts with this branch now that I've landed #4 (and thanks for that one!) I'm going to close this PR.  If there are other bits of this branch that you still want to pursue, can you please open separate PRs for them?  Again, thanks for all your great work on this code.
